collect_app/src/test/java/org/odk/collect/android/widgets/GeoPointWidgetTest.java

```diff
@@ -1,85 +1,301 @@
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearanceAndAnswer(MAPS, new StringData(answer)));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.view_change_location)));
+    }
+
+    @Test
+    public void whenWidgetHasAnswer_buttonShowsCorrectText() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.change_location)));
+    }
+
+    @Test
+    public void whenWidgetHasNullAsAnswer_buttonShowsCorrectText() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.get_point)));
+    }
+
+    @Test
+    public void whenPermissionIsNotGranted_buttonShouldNotLaunchAnyIntent() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+        stubLocationPermissions(widget, false);
+        widget.startGeoButton.performClick();
+        Intent startedIntent = shadowOf(widgetTestActivity()).getNextStartedActivity();
+
+        assertNull(startedIntent);
+    }
+
+    @Test
+    public void whenPermissionIsGranted_buttonClickShouldSetTheLocationSavedOnTheMap() {
+        FormEntryPrompt prompt = promptWithAnswer(null);
+        GeoPointWidget widget = createWidget(prompt);
+        stubLocationPermissions(widget, true);
+        widget.startGeoButton.performClick();
+
+        verify(waitingForDataRegistry).waitForData(prompt.getIndex());
+    }
+
+    @Test
+    public void whenPromptDoesNotHaveAnswer_buttonShouldLaunchCorrectIntent() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+        stubLocationPermissions(widget, true);
+        widget.startGeoButton.performClick();
+
+        Intent startedIntent = shadowOf(widgetTestActivity()).getNextStartedActivity();
+        Bundle bundle = startedIntent.getExtras();
+
+        assertThat(startedIntent.getComponent(), equalTo(new ComponentName(widgetTestActivity(), GeoPointActivity.class)));
+        assertBundleArgumentEquals(bundle, null, DEFAULT_LOCATION_ACCURACY, false, true);
+    }
+
+    @Test
+    public void whenPromptHasAnswer_buttonShouldLaunchCorrectIntent() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
+        stubLocationPermissions(widget, true);
+        widget.startGeoButton.performClick();
+
+        Intent startedIntent = shadowOf(widgetTestActivity()).getNextStartedActivity();
+        Bundle bundle = startedIntent.getExtras();
+
+        assertThat(startedIntent.getComponent(), equalTo(new ComponentName(widgetTestActivity(), GeoPointActivity.class)));
+        assertBundleArgumentEquals(bundle, GeoWidgetUtils.getLocationParamsFromStringAnswer(answer), DEFAULT_LOCATION_ACCURACY, false, true);
```

Comment:
Would be good to also set a non-default accuracy threshhold in the form entry prompt and verify that goes through to the intent.

Reply:
Ohh totally forgot that!